subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -343,56 +362,71 @@ public int size() {
         }
+        return instance;
     }
 
     private Object doGet(Type serviceType) {
-        synchronized (lock) {
+        try {
+            newRequestInProgress();
             ServiceProvider provider = providerCache.get(serviceType);
             if (provider == null) {
                 provider = find(serviceType, allServices);
-                providerCache.put(serviceType, provider);
+                if (provider == null) {
+                    provider = NOT_FOUND;
+                }
+                providerCache.putIfAbsent(serviceType, provider);
             }
-            return provider == null ? null : provider.get();
+            return provider.get();
+        } finally {
+            requestFinished();
         }
     }
 
-    public <T> Factory<T> getFactory(Class<T> type) {
-        synchronized (lock) {
-            noLongerMutable();
-            if (closed) {
-                throw new IllegalStateException(String.format("Cannot locate factory for objects of type %s, as %s has been closed.", format(type), getDisplayName()));
-            }
-            assertValidServiceType(type);
-            Factory<T> factory = doGetFactory(type);
-            if (factory == null) {
-                throw new UnknownServiceException(type, String.format("No factory for objects of type %s available in %s.", format(type), getDisplayName()));
+    private void newRequestInProgress() {
+
+        noLongerMutable();
+        if (state == State.CLOSED) {
+            throw new IllegalStateException(String.format("%s has been closed.", getDisplayName()));
+        }
+        inProgress.incrementAndGet();
+    }
+
+    private void requestFinished() {
+        boolean noMoreRequests = inProgress.decrementAndGet() == 0;
+        if (noMoreRequests && state == State.CLOSING) {
+            synchronized (stopLock) {
+                stopLock.notify();
             }
-            return factory;
         }
     }
 
+    public <T> Factory<T> getFactory(Class<T> type) {
+        assertValidServiceType(type);
+        Factory<T> factory = doGetFactory(type);
+        if (factory == null) {
+            throw new UnknownServiceException(type, String.format("No factory for objects of type %s available in %s.", format(type), getDisplayName()));
+        }
+        return factory;
+    }
+
     private <T> Factory<T> doGetFactory(Class<T> type) {
-        synchronized (lock) {
+        try {
+            newRequestInProgress();
             ServiceProvider provider = allServices.getFactory(type);
             return provider == null ? null : (Factory<T>) provider.get();
+        } finally {
+            requestFinished();
         }
     }
 
     private void noLongerMutable() {
-        if (mutable) {
-            mutable = false;
+        if (state == State.INIT) {
+            state = State.IN_USE;
```

Comment:
alright I had missed this one. It the only one that isn't synchronized, AFAICS. Good to go.